internals-js/src/operations.ts

```diff
@@ -2558,74 +2543,174 @@ abstract class AbstractSelection<TElement extends OperationElement, TIsLeaf exte
   }
 }
 
-function selectionSetsDoMerge(s1: SelectionSet, s2: SelectionSet): boolean {
-  const byResponseName1 = s1.fieldsByResponseName();
-  const byResponseName2 = s2.fieldsByResponseName();
+class FieldsConflictValidator {
+  private usedSpreadTrimmedPartAtLevel?: FieldsConflictValidator[];
+
+  private constructor(
+    private readonly byResponseName: Map<string, Map<Field, FieldsConflictValidator | null>>,
+  ) {
+  }
+
+  static build(s: SelectionSet): FieldsConflictValidator {
+    return FieldsConflictValidator.forLevel(s.fieldsInSet());
+  }
+
+  private static forLevel(level: CollectedFieldsInSet): FieldsConflictValidator {
+    const atLevel = new Map<string, Map<Field, CollectedFieldsInSet | null>>();
 
-  for (const [responseName, selections1] of byResponseName1.entries()) {
-    const selections2 = byResponseName2.get(responseName);
-    if (!selections2) {
-      // No possible conflict on this response name to check.
-      continue;
+    for (const { field } of level) {
+      const responseName = field.element.responseName();
+      let atResponseName = atLevel.get(responseName);
+      if (!atResponseName) {
+        atResponseName = new Map<Field, CollectedFieldsInSet>();
+        atLevel.set(responseName, atResponseName);
+      }
+      if (field.selectionSet) {
+        let forField = atResponseName.get(field.element) ?? [];
```

Comment:
> E.g. can there be duplicate `FieldSelection`s in the given level?

First, I'll note that the key used is `field.element`, so the question is about duplicate `Field`. Now, I'm not sure this can happen with the current code, but in theory there is no real reason why this couldn't happen, and this could easily change through some optimisations. Basically, `Field` object are just a particular field name/parent type plus potential arguments, and you can certainly "select" the same field multiple times at the same level, so while the current code probably currently always have a new `Field` object in practice, it wouldn't be crazy to do some interning of those objects in particular cases or something. Long story short, I don't want the code to be too fragile here.

>  if so, then we probably shouldn't set the value to null in the else block below

Pretty sure this is fine, because `field.selectionSet` is equivalent to saying that the underlying field type is a "leaf type" so any other selection with the same `Field` element is guaranteed to not have a selection set too. Meaning that there could already be an element for `field.element` in that else, but it is guarantee to be `null` so overriding it doesn't matter.

But I've added comments in the code for all of this as those are legitimate questions one may be wondering about on reading this.